compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractLineNumberTest.kt

```diff
@@ -57,102 +53,114 @@ abstract class AbstractLineNumberTest : CodegenTestCase() {
-
-        private fun createLineNumberDeclaration() =
-            CodegenTestCase.TestFile(
-                "$LINE_NUMBER_FUN.kt",
-                "package test;\n\npublic fun $LINE_NUMBER_FUN(): Int = 0\n"
-            )
+    protected open fun compareCustom(psiFile: KtFile, wholeFile: File) {
+        val actualLineNumbers = extractActualLineNumbersFromBytecode(classFileFactory, false)
+        val text = psiFile.text
+        val newFileText = text.substring(0 until Regex("// \\d+").find(text)!!.range.first) +
+                getActualLineNumbersAsString(actualLineNumbers)
+        KotlinTestUtils.assertEqualsToFile(wholeFile, newFileText)
+    }
 
-        private fun getActualLineNumbersAsString(lines: List<String>) =
-            lines.joinToString(" ", "// ")
+    protected fun extractActualLineNumbersFromBytecode(factory: ClassFileFactory, testFunInvoke: Boolean) =
+        factory.getClassFiles().flatMap { outputFile ->
+            val cr = ClassReader(outputFile.asByteArray())
+            if (testFunInvoke) readTestFunLineNumbers(cr) else readAllLineNumbers(cr)
+        }
 
-        private fun extractActualLineNumbersFromBytecode(factory: ClassFileFactory, testFunInvoke: Boolean) =
-            factory.getClassFiles().flatMap { outputFile ->
-                val cr = ClassReader(outputFile.asByteArray())
-                if (testFunInvoke) readTestFunLineNumbers(cr) else readAllLineNumbers(cr)
-            }
+    protected open fun readTestFunLineNumbers(cr: ClassReader): List<String> {
+        val labels = arrayListOf<Label>()
+        val labels2LineNumbers = HashMap<Label, String>()
+
+        val visitor = object : ClassVisitor(Opcodes.ASM5) {
+            override fun visitMethod(
+                access: Int,
+                name: String,
+                desc: String,
+                signature: String?,
+                exceptions: Array<String>?
+            ): MethodVisitor {
+                return object : MethodVisitor(Opcodes.ASM5) {
+                    private var lastLabel: Label? = null
+                    private var lastLine = -1
+
+                    override fun visitMethodInsn(opcode: Int, owner: String, name: String, desc: String, itf: Boolean) {
+                        if (LINE_NUMBER_FUN == name) {
+                            labels.add(lastLabel ?: error("A function call with no preceding label"))
+                        }
+                    }
 
-        private fun extractSelectedLineNumbersFromSource(file: KtFile): List<String> {
-            val fileContent = file.text
-            val lineNumbers = arrayListOf<String>()
-            val lines = StringUtil.convertLineSeparators(fileContent).split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
+                    override fun visitLabel(label: Label) {
+                        if (lastLabel != null && !labels2LineNumbers.containsKey(lastLabel) && lastLine >= 0) {
```

Comment:
Is it ir specific? Than please move condition (maybe: fun shouldKeepInheritedLines(...)) to ir part

Reply:
Moved